src/main/java/pl/km/exercise261/RecipeController.java

```diff
@@ -0,0 +1,209 @@
+    }
+
+    @GetMapping("/przepis/{id}/usun")
+    public String deleteRecipeById(@PathVariable Long id) {
+        Optional<Recipe> recipeOptional = recipeService.getById(id);
+        if (recipeOptional.isPresent()) {
+            recipeService.deleteById(id);
+        }
+        return "redirect:/";
+    }
+
+    @GetMapping("/nowy")
+    public String addNewRecipe(Model model) {
+        Recipe recipe = new Recipe();
+        List<Ingredient> ingredients = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            ingredients.add(new Ingredient());
+        }
+        recipe.setIngredients(ingredients);
+        model.addAttribute("recipe", recipe);
+        model.addAttribute("pictureName", PICTURE_FOR_RECIPE_EDIT_AND_UPDATE);
+        model.addAttribute("title", TITLE_FOR_NEW_RECIPE);
+        return "form";
+    }
+
+    @PostMapping("/przepis/dodajskladnik")
+    public String addNewIngredient(@ModelAttribute("recipe") Recipe recipe,
+                                   Model model) {
+//        List<RecipeCategory> categories = recipeCategoryRepository.findAll();
+//        model.addAttribute("categories", categories);
+        recipe.addIngredient(new Ingredient());
+        model.addAttribute("recipe", recipe);
+        String title;
+        if (recipe.getId() != null) {
+            title = "Edycja przepisu: " + recipe.getName();
+        } else {
+            title = ApplicationDisplayTemplates.TITLE_FOR_NEW_RECIPE;
+        }
+        model.addAttribute("title", title);
+        model.addAttribute("pictureName", PICTURE_FOR_RECIPE_EDIT_AND_UPDATE);
+        return "form";
+    }
+
+    @PostMapping("/przepis/usunskladnik/{index}")
+    public String removeIngredientByIndex(@ModelAttribute("recipe") Recipe recipe,
+                                          @PathVariable int index,
+                                          Model model) {
+        recipe.removeIngredientByIndex(index);
+        model.addAttribute("recipe", recipe);
+        String title;
+        if (recipe.getId() != null) {
+            title = "Edycja przepisu: " + recipe.getName();
+        } else {
+            title = ApplicationDisplayTemplates.TITLE_FOR_NEW_RECIPE;
+        }
+        model.addAttribute("title", title);
+        model.addAttribute("pictureName", PICTURE_FOR_RECIPE_EDIT_AND_UPDATE);
+        return "form";
+    }
+
+    @PostMapping("/zapisz")
+    public String saveRecipe(
+            @ModelAttribute("recipe") Recipe recipe,
+            @RequestParam(value = "inputCategories", required = false) Long[] inputCategories) {
+        if (inputCategories != null) {
+            for (int i = 0; i < inputCategories.length; i++) {
+                if (recipeCategoryRepository.findById(inputCategories[i]).isPresent()) {
+                    RecipeCategory category = recipeCategoryRepository.findById(inputCategories[i]).orElseThrow();
+                    recipe.addCategory(category);
+                }
+            }
+        }
```

Comment:
Jeżeli masz wprowadzoną warstwę usług, to w kontrolerze nie powinienes odwoływać się bezpośrednio do repozytoriów.
Tutaj dodatkowo możesz wykorzystać magię springa do automatycznego ustawiania kategorii - pokazane w code review